### ISSUE TITLE
S3#bucketExists and S3#objectExists

### ISSUE THREAD

**Issue by jhovell** [aws/aws-sdk-java-v2#392]
## Expected Behavior
I would expect feature parity for old 1.x methods such as `AmazonS3.doesObjectExist(bucket, key).` With the async client / `CompletableFuture` it's more than a minor annoyance as the result of `S3AsyncClient.headObject(...)` is to throw a NoSuchKeyException which would logically be handled in an `exceptionally(throwable -> ...)` stage but it is difficult to handle control flow such as letting other exceptions propagate. 

## Your Environment
* AWS Java SDK version used: 2.0.0-preview-7
* JDK version used: Oracle Java(TM) SE Runtime Environment (build 9+181)
Java HotSpot(TM) 64-Bit Server VM (build 9+181, mixed mode)
* Operating System and version: OSX

**Comment by millems** [comment]
+1, there needs to be more high-level operations on the S3 client. Eg. `doesBucketExist`, `doesObjectExist`, etc. 

Right now, the raw S3 operations require using exceptions for non-exceptional cases.

**Comment by redbaron** [comment]
most of the time checking for file/bucket existence buys you nothing, by the time you act on it  it might already change it's state (be deleted, be created) so unwanted state exception need to be correctly handled anyway ,  so whats the point of checking then :)

**Comment by aiba** [comment]
> most of the time checking for file/bucket existence buys you nothing, by the time you act on it it might already change it's state (be deleted, be created) 

There are plenty of systems that simply never delete s3 objects, so once you check if it exists, that won't ever change.

Also in Java I believe it's generally frowned upon to throw/catch Exceptions for situations which are expected to happen.

I think it would be useful to have a doesObjectExist method.

**Comment by anmolgarg552** [comment]
This issue is still open, so I guess this feature isn't added as of now. Correct?

I was already using the doesObjectExists in the 'aws-java-sdk-s3'. I was going to the non-blocking method but since this method doesn't exist in the java 2.x, it's a spoiler. :/

**Comment by millems** [comment]
To be clear: you can use the headObject method, but it's just a bit messy because you need to catch the exception if it doesn't exist.

**Comment by laymain** [comment]
Throwing and catching exceptions hundreds of thousands of times has a performance cost that would be good to avoid, any progress on this issue?

**Comment by yehudasag** [comment]
the exception is not thrown any way. is there a way to know if a bucket exist??

**Comment by JoshMcCullough** [comment]
Agreed with everyone above who says we need this in the Java SDK.

**Comment by fatih-celonis** [comment]
Yes, I also think that controlling the flow with the exception doesn't make any sense as it's an expensive operation and rather should be an "exception". +1 for having this method.

**Comment by RanVaknin** [comment]
Hi everyone,
Thanks for your patience on this one. We wanted to get a team wide agreement on this before making a decision.

After prototyping a few designs, we decided not to add `doesObjectExist` and `doesBucketExist` as built in convenience methods. Two things drove that:

1. The v1 client was handwritten, so it easily added these convenience utilities as first party methods on the S3 client  (`s3.doesObjectExist()`). The v2 clients are code generated from service models, so offering the same first party call on the client means wiring custom behavior through the code generation layer and maintaining it going forward. We looked at alternatives that avoid the codegen path, but none of them met our bar, either not discoverable, or not intuitive.

2. Error handling does not have one correct answer. A `headObject` or `headBucket` can return 403 instead of 404 depending on the caller's permissions, so the SDK cannot reliably treat a 403 as a clean "does not exist." Any SDK that offers these methods has to pick an interpretation of that 403, and no single choice is right for every permission setup. Rather than bake our own interpretation into the client, we decided that leaving error handling to the caller is the better design choice.

As a workaround, I added a [code example](https://github.com/awsdocs/aws-doc-sdk-examples/pull/7920) for a convenience method that does what v1 did. We know this is not the outcome some of you hoped for, but we hope the example unblocks you.

I'm going to close the issue, but feel free to leave a comment if you have any other questions or concerns.

All the best,
Ran~
